### PR TITLE
[ATO 1080] create product documentation for markers

### PR DESCRIPTION
**Pull request by vcidst**
**Proposed changes**:
- Add product documentation for Realtime Markers as per [ATO-1080](https://rasahq.atlassian.net/browse/ATO-1080)

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

[ATO-1080]: https://rasahq.atlassian.net/browse/ATO-1080?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/docs/monitoring/analytics/data-structure-reference.mdx [hunk 1]
```diff
@@ -781,6 +781,120 @@ Time of the last update of the slot in this session. The timestamp is a UTC.
 - Type: `DateTime`
 - Example: `2022-06-21 02:15:49.326936`
 
+## rasa_patterns
+
+Patterns are marker definitions that have been received from Rasa. This table is called `patterns`
+to distinguish them from extracted markers which are stored in `rasa_markers` table. It 
+stores the configuration of markers (which can be thought of as pattern
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/data-structure-reference.mdx:788]: as **a** pattern of conversational events :)

File: docs/docs/monitoring/analytics/data-structure-reference.mdx [hunk 2]
```diff
@@ -781,6 +781,120 @@ Time of the last update of the slot in this session. The timestamp is a UTC.
...
+
+Time of the last update of the pattern in this session. The timestamp is a UTC.
+
+- Type: `DateTime`
+- Example: `2022-06-21 02:15:49.326936`
+
+## rasa_markers
+
+Markers are the extracted markers from the conversations. Each row in this table corresponds
+to a conversation where the marker was applied
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/data-structure-reference.mdx:842]: is it 'applied' or 'detected' ..? Not sure.. This sentence is a lil bit unclear to me. Does each row correspond to one conversation where the marker was detected?

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+Please consult the [Markers](../../markers.mdx/#defining-markers) section of Rasa documentation
+for details about defining Markers.
+
+## Enable Real-time Processing
+
+To enable real-time analysis of Markers in your Rasa project, you need to
+upload the markers to Analytics Data Pipeline with the `rasa marker upload`
+command. This command sends the marker configuration YAML and starts the
+processing of Markers in the Analytics Data Pipeline. To get started,
+make sure you have the latest version of Rasa installed. Then, open your command line
```
Comment (radovanZRasa) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:39]: We should mention what is the minimum version of Rasa-Pro which supports real-time-markers.

File: docs/docs/monitoring/analytics/data-structure-reference.mdx (hunk 1 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/data-structure-reference.mdx:788]: Oh right, I'll fix it

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import useBaseUrl from "@docusaurus/useBaseUrl";
+import RasaProLabel from "@theme/RasaProLabel";
+import RasaProBanner from "@theme/RasaProBanner";
+import { Grid } from "@theme/Grid";
+
+<RasaProLabel />
+
+<RasaProBanner />
+
+In Rasa, [Markers](../../markers.mdx) are a powerful feature that allows
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:21]: In Rasa Pro, Markers is a powerful feature that allows ...

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 3 above)
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:39]: latest (would it be more true to say 3.6 and higher?) version of Rasa Pro installed.

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+This command should be run whenever there is a change in the marker
+configuration file. The changes might include addition of new markers,
+changing an existing marker or removing an existing marker. This command
+will upload the configuration file to the Analytics Data Pipeline
+
+:::note
+
+This command **uploads** the marker configurations to the data pipeline. The pipeline assusmes the configuration
+file as the source of truth and only processes the markers defined in it. If you remove a marker from this file,
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:58]: file is the source of truth

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+Visit our [CLI page](./command-line-interface.mdx#rasa-marker-upload)
+for more information about the command line arguments available.
+
+## How are Markers processed?
+
+Once the marker definitions are uploaded to the Analytics Data Pipeline, they
+are stored in the `rasa_patterns` table. As the Kafka Consumer starts receiving events
+from the Rasa Assistant, it starts analyzing them for markers. The Pipeline 
+processes all the events from the Kafka Event Broker and identifies points of interest
+in the conversation where the marker can be applied. The extracted markers are stored
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:74]: would it be better to say: identifies points of interests in the conversation that matches the marker ? what do you think?

File: docs/docs/monitoring/analytics/data-structure-reference.mdx (hunk 2 above)
Comment (andreaklmn) [reply docs/docs/monitoring/analytics/data-structure-reference.mdx:842]: okay,  I don't have a strong opinion on this,

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 3 above)
Comment (andreaklmn) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:39]: Yup, I would say here: 
To get started, please make sure you have Rasa Pro 3.6 or higher installed.

File: docs/docs/monitoring/analytics/data-structure-reference.mdx (hunk 1 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/data-structure-reference.mdx:788]: Fixed, thanks!

File: docs/docs/monitoring/analytics/data-structure-reference.mdx (hunk 2 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/data-structure-reference.mdx:842]: The existing rasa documentation uses the collocation "markers applied" so I kept it that way, [here's the section](https://rasa.com/docs/rasa/markers/#extracted-markers).

> Does each row correspond to one conversation where the marker was detected?
Each row corresponds to one marker, not specifically to one conversation. I'll rephrase the sentence

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 3 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:39]: I've added a info section at the top of this page to specify its available in Rasa Pro from 3.6 onwards

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 5 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:58]: Fixed this

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+optional `-d` argument.
+
+This command should be run whenever there is a change in the marker
+configuration file. The changes might include addition of new markers,
+changing an existing marker or removing an existing marker. This command
+will upload the configuration file to the Analytics Data Pipeline
+
+:::note
+
+This command **uploads** the marker configurations to the data pipeline. The pipeline assusmes the configuration
```
[the remark this answers is not in the document]
Comment (vcidst) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:57]: Yes, fixed

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 6 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:74]: Yep, that sounds better. I've changed it

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+```
+rasa markers upload --config=<path-to-config-file> -d=<path-to-domain-file> -rasa-pro-services-url=<url>
+```
+
+By default, this command validates the marker configuration file against
+your bot's domain.yml file. To specify a different domain file, use the
+optional `-d` argument.
+
+This command should be run whenever there is a change in the marker
```
Comment (rgstephens) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:53]: How do you stop forwarding real-time markers?

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+from the Rasa Assistant, it starts analyzing them for markers. The Pipeline 
+processes all the events from the Kafka Event Broker and identifies points of 
+interests in the conversation that matches the marker. The extracted markers are stored
+in the data warehouse.
+
+The evaluation of Markers in the Pipeline is similar to the `rasa evaluate markers`
+command which can be used to process markers from the conversations in Tracker Store.
+Read more about it [here](../../markers.mdx/#extracting-markers)
+
+Extracted markers are added to the `rasa_markers` table in the data warehouse
```
Comment (rgstephens) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:83]: Can you add an example query and graph for markers on the Example queries page?

**Review by rgstephens — CHANGES_REQUESTED** [review]

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 8 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:53]: I'm sorry I didn't understand your question. Did you mean to ask how does one stop uploading the marker-configurations to rasa-pro-services? The marker-configurations are only forwarded to rasa-pro-services (referred as Analytics Data Pipeline in the documentation) when this command is run.

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 9 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:83]: I can add an example query and graph to visualise `num_preceeding_user_turns` for a marker that tracks when user has completed the form. It'll visualise the number of turns users took to complete the form and can help give insight if the users are struggling to complete the form. Will something like that be helpful? It's only one of the many use cases of this feature

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+:::
+
+### Configuring the CLI command
+
+Visit our [CLI page](./command-line-interface.mdx#rasa-marker-upload)
+for more information about the command line arguments available.
+
+## How are Markers processed?
+
+Once the marker definitions are uploaded to the Analytics Data Pipeline, they
```
Comment (rgstephens) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:72]: What's a "marker definition"? Above, it says "This command uploads the marker configurations to the data pipeline." and here it's talking about the "marker definitions are uploaded"?

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 8 above)
Comment (rgstephens) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:53]: I have the impression that this command starts "real time processing" of markers and that it kicks off real-time forwarding of markers to the analytics pipeline by Rasa.

I must be confused about what happens when you run this command.

Does the command do a one time upload of markers to the analytics and then exit?  Does it run and forward markers until you manually stop the command?

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 9 above)
Comment (rgstephens) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:83]: A form related example would be really good. When Alana and I recently wrote the blog post on this subject, we really needed a completion rate example. For a given form or story, how many people started the form and how many completed it successfully.

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 8 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:53]: This command uploads the marker-configuration(aka patterns) files and then exits. Rasa Analytics stores these patterns in a database table called `rasa_patterns`. To remove any pattern, you can remove them from the YAML file and run `rasa markers upload` again.

Rasa Analytics will then process all events from Kafka and extract any markers that it founds in the stream of events. The markers are extracted based on the patterns in `rasa_patterns` table. Extracted markers are stored in `rasa_markers` table

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 9 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:83]: > For a given form or story, how many people started the form and how many completed it successfully.

I'll add this example to this documentation

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 10 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:72]: I agree that's a bit confusing. Marker definition or marker configuration refers here to the YAML files that describe the marker in the following manner,

```
marker_cheer_up_success:
  description: "Sequence of intent and action that define the marker_cheer_up_success"
  seq:
    - intent: mood_unhappy
    - action: utter_cheer_up
    - action: utter_did_that_help
    - intent: affirm
```

This information is used to identify when these event sequences occur in the conversation. Or, when `marker_cheer_up_success` is extracted from a conversation. To differentiate these, I have referred the YAML files as "marker-configuration" and the markers found in the conversations as "extracted markers". 

Do you have any suggestions for making this distinction more clear and simplified?

File: docs/docs/monitoring/analytics/realtime-markers.mdx (hunk 9 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/realtime-markers.mdx:85]: @rgstephens With a release schedule of next week, we're short on time to add this example at the moment but I've created a ticket at https://rasahq.atlassian.net/browse/ATO-1227 for this and we'll add this example in a micro release

File: docs/docs/monitoring/analytics/data-structure-reference.mdx [hunk 11]
```diff
@@ -781,6 +781,120 @@ Time of the last update of the slot in this session. The timestamp is a UTC.
...
+#### `updated_at` update date time
+
+Time of the last update of the pattern in this session. The timestamp is a UTC.
+
+- Type: `DateTime`
+- Example: `2022-06-21 02:15:49.326936`
+
+### rasa_markers
+
+Markers are the extracted markers from the conversations. Each row in this table corresponds
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/data-structure-reference.mdx:841]: This sentence reads a bit weird:) Markers are the extracted markers from the conversations.

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+are stored in the `rasa_patterns` table. As the Kafka Consumer starts receiving events
+from the Rasa Assistant, it starts analyzing them for markers. The Pipeline 
+processes all the events from the Kafka Event Broker and identifies points of 
+interests in the conversation that matches the marker. The extracted markers are stored
+in the data warehouse.
+
+The evaluation of Markers in the Pipeline is similar to the `rasa evaluate markers`
+command which can be used to process markers from the conversations in Tracker Store.
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:82]:
```suggestion
command which can be used to process Markers from the conversations in Tracker Store.
```

File: docs/docs/command-line-interface.mdx [hunk 13]
```diff
@@ -661,6 +662,68 @@ Python Logging Options:
...
+  -vv, --debug          Print lots of debugging statements. Sets logging level
+                        to DEBUG. (default: None)
+  --quiet               Be quiet! Sets logging level to WARNING. (default:
+                        None)
+  --logging-config-file LOGGING_CONFIG_FILE
+                        If set, the name of the logging configuration file
+                        will be set to the given name. (default: None)
+
+Description:
+  The `rasa markers upload` command allows you to upload Markers to the Rasa Pro Services. Markers are custom conversational events that provide additional context for analysis and insights generation. By uploading Markers, you can enable real-time analysis and enhance the performance of your Rasa Assistant.
```
Comment (andreaklmn) [inline docs/docs/command-line-interface.mdx:719]:
```suggestion
  The `rasa markers upload` command allows you to upload markers to the Rasa Pro Services. Markers are custom conversational events that provide additional context for analysis and insights generation. By uploading markers, you can enable real-time analysis and enhance the performance of your Rasa Assistant.
```

File: docs/docs/markers.mdx [hunk 14]
```diff
@@ -14,6 +14,12 @@ This feature is currently experimental and might change or be removed in the fut
 
 :::
 
+:::danger
+
+In the upcoming release version 3.7 of Rasa Open Source, we’re removing this experimental feature. For documentation on the Markers feature in Rasa Pro, please [click here](./monitoring/analytics/realtime-markers.mdx)
```
Comment (andreaklmn) [inline docs/docs/markers.mdx:19]:
```suggestion
In the upcoming release version 3.7 of Rasa Open Source, we’re removing this experimental feature. For documentation on the markers feature in Rasa Pro, please [click here](./monitoring/analytics/realtime-markers.mdx)
```

File: docs/docs/markers.mdx [hunk 15]
```diff
@@ -173,6 +179,12 @@ You cannot reuse an existing marker name in the definition of another marker.
 
 ## Extracting Markers
 
+:::info
+
+Rasa Pro supports real-time processing of Markers, Read more about [Real-Time Markers](monitoring/analytics/realtime-markers.mdx)
```
Comment (andreaklmn) [inline docs/docs/markers.mdx:184]:
```suggestion
Rasa Pro supports real-time processing of markers, Read more about [Real-Time Markers](monitoring/analytics/realtime-markers.mdx)
```

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 16]
```diff
@@ -0,0 +1,89 @@
+---
+id: realtime-markers
+sidebar_label: Real-Time Markers
+title: Real-Time Analysis of Markers
+description:
+abstract: Process Markers real-time from Rasa
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:6]:
```suggestion
abstract: Process markers real-time from Rasa
```

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+<RasaProLabel />
+
+:::info Available in Rasa Pro from 3.6.0
+:::
+
+<RasaProBanner />
+
+In Rasa Pro, [Markers](../../markers.mdx) are a powerful feature that allows
+you to track and extract custom conversational events. With the Analytics
+Data Pipeline, you can now process these Markers in real-time, enabling
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:26]:
```suggestion
Data Pipeline, you can now process these markers in real-time, enabling
```

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+you to gain valuable insights and enhance the performance of your Rasa
+Assistant. In this guide, we'll explore how to leverage real-time
+analysis of Markers to track solution and abandonment rates in your conversations.
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:29]:
```suggestion
analysis of markers to track solution and abandonment rates in your conversations.
```

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
+![flow of information with realtime markers](/img/analytics/realtime-markers.png)
+
+## Defining Markers
+
+Please consult the [Markers](../../markers.mdx/#defining-markers) section of Rasa documentation
+for details about defining Markers.
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:36]:
```suggestion
for details about defining markers.
```

File: docs/docs/monitoring/analytics/realtime-markers.mdx [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+
+## Enable Real-time Processing
+
+To enable real-time analysis of Markers in your Rasa project, you need to
```
Comment (andreaklmn) [inline docs/docs/monitoring/analytics/realtime-markers.mdx:40]:
```suggestion
To enable real-time analysis of markers in your Rasa project, you need to
```

**Review by andreaklmn — APPROVED** [review]
I ended up making Markers lower case everywhere to be a bit more consistent

File: docs/docs/monitoring/analytics/data-structure-reference.mdx (hunk 11 above)
Comment (vcidst) [reply docs/docs/monitoring/analytics/data-structure-reference.mdx:841]: Fixed it

**Comment by vcidst** [comment]
Hey @rgstephens, I've addressed the sections you pointed out and expanded on them. Since today is the release day, I'll go ahead and merge this PR. However, I still appreciate your input, so if you have any further thoughts or suggestions, please let me know. I'll make sure to incorporate them into the documentation in a micro release. Thanks!

**Pull request merged**